extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java

```diff
@@ -1249,6 +1249,16 @@ public static Token fromAudience(String... audience) {
         @ConfigItem(defaultValue = "true")
         public boolean allowOpaqueTokenIntrospection = true;
 
+        /**
+         * Token customizer name.
+         *
+         * Allows to select a tenant specific token customizer.
+         * If this property is not set then a default `io.quarkus.oidc.TokenCustomizer` bean
+         * will be used if registered.
+         */
+        @ConfigItem
+        public Optional<String> tokenCustomizer = Optional.empty();
```

Comment:
Do we really want to pass a class name as part of configuration? This is not user friendly (as the IDE can't help you) and we generally try to avoid it.

Wouldn't it make more sense to just look for beans of type `TokenCustomizer`?

Reply:
Hi @geoand, totally agreed re using class names, this is though a bean name, I'll improve the docs. In Quarkus OIDC a single endpoint can support more than one provider with a multi-tenant setup, for ex:
```
quarkus.oidc.tenant1.provider=microsoft
quarkus.oidc.tenant2.provider=twitch
```  

This is just an example (Twitch PR is nearly complete). Each of these providers may have its own peculiarities in how to verify the token, so this property allows to do something like:

```
quarkus.oidc.tenant1.provider=microsoft
quarkus.oidc.tenant1.token-customizer=azure-graph-token-customizer
quarkus.oidc.tenant2.provider=twitch
quarkus.oidc.tenant1.token-customizer=twitch-token-customizer
```  

where each provider will have its own customizer like
```
@Named("azure-graph-token-customizer")
@ApplicationScoped
class AzureCustomizer implements TokenCustomizer {
}
``` 

But also this property does not have to be set. Indeed, if it won't be set and we just have a TokenCustomizer bean then it will be used for all the tenants. May be if we have N tenants representing N Keycloak realms, or just a single/default OIDC tenant.

Hope it clarifies it :-)

Reply:
Also the interface is a WIP, I might wrap JsonObject in something like `TokenHeaders` etc

Reply:
>  this is though a bean name

That is not something we rely on almost anywhere. It's generally bad practice to use `@Named`.

Reply:
I know this is a WIP thing, but my point is that we should be relying on bean types and qualifiers is necessary - not bean names.

Reply:
Why not just use a CDI qualifier like `@Tenant("tenant1")`? That way users only need to know one string per configured tenant.
I see no good reason for them to have to configure another String in configuration.

Reply:
@geoand 

>Why not just use a CDI qualifier like @Tenant("tenant1")?

It looks fine, though I recall we had problems with supporting `@OidcClient("client1")`  - though Michal Vavrik did it in the end. May be it was related to `RequestScoped` aspects, here a pattern is different, we just need to find `@TenantId` whose value is set to a given tenant id...

That said, it has its limitations now that I think about it, say we have more than one Azure tenant supported, we have users supporting N Keycloak realms (essentially one realm is one tenant), so we can have many Azure tenants, I won't be able to bind a single class to more than one tenant with this approach.  

In that case though we can just have a single global bean without `@TenantId` which will be picked up by every tenant - but here it breaks if alongside these say 2 Azure tenants managing Azure Graph tokens, 2 tenants dealing with Azure SomeOtherService tokens are introduced so then a global bean idea meant to cover Azure Graph tokens fails. This is though a kind of niche case

>I see no good reason for them to have to configure another String in configuration.

I'm conscious now of adding more properties :-). And `@Tenant("azure-graph")` would cover many practical cases, the whole customization of tokens is a fairly rare case on its own. However, I'd like to avoid introducing any constraints on a multi-tenant setup. Right now I'm fine with dropping the idea of depending on `@Named` with a `token-customizer` property - but I'll introduce a fallback to `@Named` if users will ever come back sayng - we can't cover all tenants with it.

Yeah, I think it could a reasonable strategy for a multi-tenant setup as we may want to support other beans:
1 Try @Tenant("tenant-id")
2 (won't be available yet) - try a property/@Named as a fallback
3 Check a global bean

I'll continue with 1 and 3 for now...

Cheers

Reply:
Makes sense

Reply:
Hi @geoand 
I've worked on finalizing this PR for the last few days, and I did end up with combining both named and `@Tenant` look up strategies. 
As far as `@Tenant("tenant-id")` is concerned, I've been having some mixed feelings about it, given that for example an Azure customizer is a feature while here we refer to a configuration property identifying a tenant to which this feature applies and such property names might vary - it is not exactly a very  type-safe resolution. And it does not work if, as in case of this PR, Quarkus ships a customizer itself - I don't want users having to extend it or type that kind of code manually for the Azure case.

However, what I `do like` about `@Tenant` is that it introduces a very effective mechanism to associate more than one custom feature with a single tenant. For example, if we have 3 features, then instead of doing 3 configuration properties, we just do:
```
@Tenant("tenant")
class MyClass impements TokenCustomer, Feature2, Feature3 {}
```

which is why I did introduce `@Tenant` annotation - it will prove very valuable for OIDC. As far as this PR is concerned - users are encouraged to use `@Tenant` to register custom customizers, and `integration-tests/oidc-wiremock` confirms it works with `TokenCustomer`.

I hope it is a reasonable compromise as far as the bean resolution logic is concerned.

Reply:
> And it does not work if, as in case of this PR, Quarkus ships a customizer itself - I don't want users having to extend it or type that kind of code manually for the Azure case.

This is not true - you can support multiple customizers and have them use an order, see for example what we do with `ObjectMapperCustomizer`.

Reply:
> I hope it is a reasonable compromise as far as the bean resolution logic is concerned.

It's up to you of course, but I am still against requiring users to wire things up with `@Named` and a configuration property.

What's more, `@Named` is global, so user could potentially run into issues if they have beans of the same name

Reply:
Hi @geoand, before commenting below: your feedback is very much appreciated, it is not that I'm saying `I'll do it my way no matter what`, I do my best to try to take your feedback on board :-) and look forward to you reviewing, which is why custom customizers are tested with `@Tenant("id")` and docs recommend `@Tenant("id")` (even though, for this PR's case), but nonetheless, I'm finding `@Tenant("someid")` having its own constraints - and I don't want OIDC multi-tenant users be constrained - which is why `@Tenant` is part of the solution but not the whole solution.

>This is not true - you can support multiple customizers and have them use an order, see for example what we do with ObjectMapperCustomizer

Sure. The question is, how one can refer to a token customizer shipped by OIDC in a multi-tenant setup or even a single tenant setup. Lets say we have `azure-tenant-1` : 1) we don't know if this tenant needs customization - it does in case of the issue this PR addresses, but then it does not with Azure tokens we dealt with in the users forum yesterday 2) Given `azure-tenant-1`  - how can I tell Quarkus I need this tenant to reuse the customizer shipped by Quarkus with `@Named` ? 3) Given `azure-tenant-1`, `azure-tenant-2` (and add `auth0-tenant-1` later) - how both of them can reuse the tenant shipped by Quarkus...

>What's more, @Named is global, so user could potentially run into issues if they have beans of the same name

Sure, @Named is loose, but look, we have `CredentialsProvider` API based on the `@Named` idea - I'm not saying it was the best idea, but with `CredentialsProvider` API one can have many providers at the same time, and we have never had a single issue with `@Named`. 

As far as this PR is concerned, `@Tenant("id")` is a first class citizen. `@Named` is only there to help users refer to the shipped Azure customizer (I feel it might possibly be the only real use case where such an advanced token customization is needed and offered by Quarkus  itself, 1/2 more max I'd say). 
`@Tenant("id")` will help with avoiding adding more properties  for sure, as I typed the the code example above.

Cheers

Reply:
> it is not that I'm saying I'll do it my way no matter what, I do my best to try to take your feedback on board :-)

I know :). 

We are both after the same thing, trying to create the best possible experience for Quarkus users.

> Sure. The question is, how one can refer to a token customizer shipped by OIDC in a multi-tenant setup or even a single tenant setup. Lets say we have azure-tenant-1 : 1) we don't know if this tenant needs customization - it does in case of the issue this PR addresses, but then it does not with Azure tokens we dealt with in the users forum yesterday - though to be fair the customizer tries to be selective 2) Given azure-tenant-1 - how can I tell Quarkus I need this tenant to reuse the customizer shipped by Quarkus without @Named ? 3) Given azure-tenant-1, azure-tenant-2 (and add auth0-tenant-1 later) - how both of them can reuse the tenant shipped by Quarkus...

Understood, thanks for the examples.

In that case, would it then not be possible to amend the API to pass the tenant to the customizer and have it decide? This could be a separate method invoked before the `customize` method but it's not strictly necessary.

This approach seems better to me.


> Sure, @Named is loose, but look, we have CredentialsProvider API based on the @Named idea

Which I very much dislike as well - it's very confusing and it's super easy to make a mistake.

Reply:
👍🏼